File: TicTacToe/TicTacToe.py
class TicTacToe:
    #constructorul clasei
    def __init__(self):
        self.board = [['-' for _ in range(3)] for _ in range(3)]
        self.currentPlayer = 'X';

    def display_board(self):
        print("  0 1 2")
        for i in range(3):
            print(f"{i} {' '.join(self.board[i])}")

    # row = 3
    # col = 7
    #not false = true
    # is_valid(self, 3, 7)

    # facem pentru o mutare valida
    # row = 1
    # col = 1
    def is_valid_move(self, row, col):

#                   3                 7
        if not(0 <= row <= 2 and 0 <= col <= 2):
            # TRUE
            #
            # not TRUE = FALSE
            return False

        return self.board[row][col] == '-'#daca in celula respectiva se afla '-', atunci mutarea este valida
        # altfel returneaza False

    def has_won(self):

        # verificare pe linii
        for row in self.board:
            if row.count(self.currentPlayer) == 3:
                return True

        # verificare pe coloane
        for col in range( 3 ):
            if all(self.board[row][col] == self.currentPlayer for row in range(3)):
                return True

        # verificare pe diagonala principala
        if all(self.board[i][i] == self.currentPlayer for i in range( 3 )):
            return True

        # verificare pe diagonala secundara
        #i = 0, 2
        if all(self.board[i][2-i] == self.currentPlayer for i in range(3)):
            return True

    def is_board_full(self):
        return all(cell != '-' for row in self.board for cell in row)

    def playGame(self):

        gameFinished = False

        while not gameFinished:

            self.display_board()

            print(f"Jucator {self.currentPlayer}, introduceti linia (0-2) si coloana (0-2):")

            try:
                row = int(input("Linia: "))
                col = int(input("Coloana: "))

            except ValueError:
                print("Introduceti numere valide")
                continue

            if self.is_valid_move(row, col):
                self.board[row][col] = self.currentPlayer

                if self.has_won():
                    self.display_board()
                    print(f"Jucatorul {self.currentPlayer} a castigat!")
                    gameFinished = True
                elif self.is_board_full():
                    self.display_board()
                    print("Joc la egalitate!")
                    gameFinished = True
                else:
                    self.currentPlayer = '0' if self.currentPlayer == 'X' else 'X'
            else:
                print("Mutare invalida! Incearca din nou sa muti!")

if __name__ == "__main__":
    game = TicTacToe()
    game.playGame();

File: TicTacToe/TicTacToe.cpp
#include <iostream>

using namespace std;

class TicTacToe {

    private:
      //data private
    char board[ 3 ][ 3 ],

        currentPlayer; //X ,0, X, 0

    void initBoard() {

        for(int i = 0; i < 3; ++i) {

            for(int j = 0; j < 3; ++j) {

                board[i][j] = '-';
            }
        }
    }

    void displayBoard() {

        cout << "  0 1 2\n";
        for(int i = 0; i < 3; ++i) {

            cout << i << " ";

            for(int j = 0; j < 3; ++j) {

                cout << board[i][j] << " ";
            }
            cout << "\n";
        }
    }

    bool isValidMove(int row, int col) {

        if(row < 0 || row > 2 || col < 0 || col > 2) return false;

        return board[row][col] == '-';
    }

    bool isBoardFull() {

        for(int i = 0; i < 3; ++i) {

            for(int j = 0; j < 3; ++j) {

                if(board[i][j] == '-') return false;
            }
        }
        return true;
    }

    bool hasWon() {

        // verificare pe linii:
        for(int i = 0; i < 3; ++i) {

            if(board[i][0] == currentPlayer && board[i][1] == currentPlayer && board[i][2] == currentPlayer) return true;
        }

        // verificare pe coloana
        for(int j = 0; j < 3; ++j) {

            if(board[0][j] == currentPlayer && board[1][j] == currentPlayer && board[2][j] == currentPlayer) return true;
        }

        // verificare pe cele doua diagonale
        if(board[0][0] == currentPlayer && board[1][1] == currentPlayer && board[2][2] == currentPlayer) return true;

        if(board[0][2] == currentPlayer && board[1][1] == currentPlayer && board[2][0] == currentPlayer) return true;

        return false;
    }

    public:

    // constructorul clasei
    TicTacToe() {
        currentPlayer = 'X';
        initBoard();
    }

    void playGame() {

        bool gameFinished = false;

        int row, col;

        while(!gameFinished) {

            displayBoard();

            cout << "Jucator " << currentPlayer << ", introduceti linia (0 - 2) si coloana (0 - 2): ";
            cin >> row >> col;

            // daca este valida mutarea
            if(isValidMove(row, col)) {

                board[ row ][ col ] = currentPlayer;

                if( hasWon() ) {
                    
                    displayBoard();
                    cout << "Jucatorul " << currentPlayer << " a castigat!\n";
                    gameFinished = true;

                } else if(isBoardFull()) {
                    
                    displayBoard();
                    cout << "Joc la egalitate!";
                    gameFinished = true;

                } else {
                    currentPlayer = (currentPlayer == 'X') ? '0' : 'X';
                }
            
            } else {
                cout << "Mutare invalida! Incearca din nou sa muti!";
            }
        }
    }
};

int main(int argc, char const *argv[]) {

  TicTacToe game;

  game.playGame();

  return 0;
}
